jni/events/sampleevent.cpp

```diff
@@ -182,17 +176,197 @@ void SampleEvent::setSample( AudioBuffer* sampleBuffer )
         _locked = false;
 }
 
-void SampleEvent::mixBuffer( AudioBuffer* outputBuffer, int bufferPos, int minBufferPosition, int maxBufferPosition,
+float SampleEvent::getPlaybackRate()
+{
+    return _playbackRate;
+}
+
+void SampleEvent::setPlaybackRate( float value )
+{
+    // allow only 100x slowdown and speed up
+    _playbackRate = std::max( 0.01f, std::min( 100.f, value ));
+}
+
+int SampleEvent::getEventLength()
+{
+    return ( _playbackRate == 1.f ) ? _eventLength : ( int )( _eventLength / _playbackRate );
+}
+
+int SampleEvent::getEventEnd()
+{
+    return ( _playbackRate == 1.f ) ? _eventEnd : _eventStart + getEventLength();
+}
+
+void SampleEvent::mixBuffer( AudioBuffer* outputBuffer, int bufferPosition,
+                             int minBufferPosition, int maxBufferPosition,
                              bool loopStarted, int loopOffset, bool useChannelRange )
 {
+    if ( !hasBuffer() )
+        return;
+
     // if we have a range length that is unequal to the total sample duration, read from the range
     // otherwise invoke the base mixBuffer method
 
-    if ( _useBufferRange )
-        getBufferForRange( outputBuffer, bufferPos );
-    else
-        BaseAudioEvent::mixBuffer( outputBuffer, bufferPos, minBufferPosition, maxBufferPosition,
+    if ( _useBufferRange ) {
+        getBufferForRange( outputBuffer, bufferPosition );
+        return;
+    }
+
+    if ( _playbackRate == 1.f ) {
+        // use BaseAudioEvent behaviour if no custom playback rate is requested
+        BaseAudioEvent::mixBuffer( outputBuffer, bufferPosition, minBufferPosition, maxBufferPosition,
                                    loopStarted, loopOffset, useChannelRange );
+        return;
+    }
+
+    // custom playback rate
+
```

Comment:
A big chunk of code just for doing the playback rate fixing, this is completely within the realm of _SampleEvent_ though as this should not apply to other AudioEvents (e.g. _SynthEvent_).